File: bindings/java/src/org/raml/amf/core/domain/shapes/Shape.java
package org.raml.amf.core.domain.shapes;

import org.raml.amf.core.domain.DomainModel;
import org.raml.amf.utils.Clojure;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by antoniogarrote on 15/05/2017.
 */
public class Shape extends DomainModel {

    public final static String SH_NS = "http://www.w3.org/ns/shacl#";
    public final static String SHAPES_NS = "http://raml.org/vocabularies/shapes#";


    public Shape(Object rawModel) {
        super(rawModel);
    }

    public String getId() {
        return (String) Clojure.get(this.rawModel,"@id");
    }

    public List<Shape> getInherits() {
        List<Object> inherits = this.getRdfObjects(SHAPES_NS + "inherits");
        List<Shape> acc = new ArrayList<>(inherits.size());
        for (Object e : inherits) {
            acc.add(Shape.fromRawModel(e));
        }
        return acc;
    }

    public void setInherits(List<Shape> shapes) {
        this.setRdfObject(SHAPES_NS + "inherits", shapes);
    }

    public static Shape fromRawModel(Object rawModel) {

        if (Shape.findShapeClass(rawModel, SH_NS + "NodeShape")) {
            return new NodeShape(rawModel);
        } else if (Shape.findShapeClass(rawModel, SH_NS + "PropertyShape")) {
            return new PropertyShape(rawModel);
        } else if (Shape.findShapeClass(rawModel, SHAPES_NS + "Scalar")){
            return new ScalarShape(rawModel);
        } else if (Shape.findShapeClass(rawModel, SHAPES_NS + "Array")) {
            return new ArrayShape(rawModel);
        } else if (Shape.findShapeClass(rawModel, SHAPES_NS + "FileUpload")) {
            return new  FileUploadShape(rawModel);
        } else if (Shape.findShapeClass(rawModel, SHAPES_NS + "XMLSchema")) {
            return new XMLSchemaShape(rawModel);
        } else if (Shape.findShapeClass(rawModel, SHAPES_NS + "JSONSchema")) {
            return new JSONSchemaShape(rawModel);
        } else {
            return new Shape(rawModel);
        }
    }

    protected static boolean findShapeClass(Object shape, String shapeClass) {
        List<Object> types = Shape.ensureArray(Clojure.get(shape, "@type"));
        for(Object type : types) {
            if (shapeClass.equals(type)) {
                return true;
            }
        }
        return false;
    }

    public static Shape build(String id) {
        Object acc = Clojure.emptyMap();
        acc = Clojure.set(acc, "@id", id);
        List<String> types = new ArrayList<>();
        types.add(id);
        acc = Clojure.set(acc, "@type", Clojure.list(types));
        return new Shape(acc);
    }

    protected <T> T getValue(String p) {
        List<Object> res = Shape.ensureArray(Clojure.get(this.rawModel,p));
        if (res.size() > 0) {
            Object e = res.get(0);
            e = Clojure.get(e, "@value");
            try {
                return (T) e;
            } catch (Exception ex) {
                return null;
            }
        } else {
            return null;
        }
    }

    protected <T> List<T> getValues(String p)  {
        List<Object> res = Shape.ensureArray(Clojure.get(this.rawModel, p));
        List<T> acc = new ArrayList<>(res.size());
        if (res.size() > 0) {
            for(Object x : res) {
                try {
                    acc.add((T) x);
                } catch(Exception e) {
                    // ignore
                }
            }

        }
        return acc;
    }

    protected void setValue(String p, Object x) {
        if (x == null) {
            Clojure.remove(this.rawModel, p);
        } else {
            List<Object> acc = new ArrayList<>();
            acc.add(Clojure.set(Clojure.map(), "@value", x));
            this.rawModel = Clojure.set(this.rawModel, p, Clojure.list(acc));
        }
    }

    protected void setValues(String p, Object x) {
        if (x == null) {
            this.rawModel = Clojure.remove(this.rawModel, p);
        } else {
            List<Object> xs = Shape.ensureArray(x);
            List<Object> xsp = new ArrayList(xs.size());
            for(Object e : xs) {
                xsp.add(Clojure.set(Clojure.emptyMap(), "@value", e));
            }
            this.rawModel = Clojure.set(this.rawModel, p, Clojure.list(xsp));
        }
    }

    protected void setRdfObject(String p, Object object) {
        if (object == null) {
            this.rawModel = Clojure.remove(this.rawModel, p);
        } else {
            this.rawModel = Clojure.set(this.rawModel, p, Shape.ensureArray(object));
        }
    }

    protected Object getRdfObject(String p) {
        List<Object> res = Shape.ensureArray(Clojure.get(this.rawModel,p));
        if (res.size() > 0) {
            return res.get(0);
        } else {
            return null;
        }
    }

    protected List<Object> getRdfObjects(String p) {
        return Shape.ensureArray(Clojure.get(this.rawModel, p));
    }

    protected String getRdfId(String p) {
        List<Object> res = Shape.ensureArray(Clojure.get(this.rawModel,p));
        if (res.size() > 0) {
            return (String) Clojure.get(res.get(0), "@id");
        } else {
            return null;
        }
    }

    protected void setRdfId(String p, String id) {
        if (id == null) {
            this.rawModel = Clojure.remove(this.rawModel,p);
        } else {
            Object idMap = Clojure.set(Clojure.emptyMap(), "@id", id);
            List acc = new ArrayList();
            acc.add(idMap);
            this.rawModel = Clojure.set(this.rawModel, p, Clojure.list(acc));
        }
    }

    protected static List<Object> ensureArray(Object x) {
        if (x == null) {
            return new ArrayList<>();
        } else if(x instanceof List) {
            return Clojure.toJavaList((List) x);
        } else {
            List<Object> acc = new ArrayList();
            acc.add(x);
            return acc;
        }
    }

}

File: bindings/js/src/core/domain/shapes/Shape.ts
/**
 * Created by antoniogarrote on 10/05/2017.
 */

import {DomainModel} from "../DomainModel";

export const SH_NS = "http://www.w3.org/ns/shacl#";
export const SHAPES_NS = "http://raml.org/vocabularies/shapes#";
export type XSDDataType = string;

export class Shape extends DomainModel {

    public getId(): string {
        return this.rawModel["@id"];
    }

    public getInherits(): Shape[] {
        return this.getRdfObjects(`${SHAPES_NS}inherits`).map(jsonld => Shape.fromRawModel(jsonld));
    }

    public setInherits(shapes: Shape[]) {
        this.setRdfObject(`${SHAPES_NS}inherits`, shapes);
    }

    public static build(id: string): Shape {
        return new Shape({"@id": id, "@type": [`${SH_NS}Shape`]})
    }

    public static fromRawModel(rawModel: any): Shape {
        if (this.findShapeClass(rawModel, `${SH_NS}NodeShape`)) {
            return new NodeShape(rawModel);
        } else if (this.findShapeClass(rawModel, `${SH_NS}PropertyShape`)) {
            return new PropertyShape(rawModel);
        } else if (this.findShapeClass(rawModel, `${SHAPES_NS}Scalar`)){
            return new ScalarShape(rawModel)
        } else if (this.findShapeClass(rawModel, `${SHAPES_NS}Array`)) {
            return new ArrayShape(rawModel);
        } else if (this.findShapeClass(rawModel, `${SHAPES_NS}FileUpload`)) {
            return new  FileUploadShape(rawModel);
        } else if (this.findShapeClass(rawModel, `${SHAPES_NS}XMLSchema`)) {
            return new XMLSchemaShape(rawModel);
        } else if (this.findShapeClass(rawModel, `${SHAPES_NS}JSONSchema`)) {
            return new JSONSchemaShape(rawModel);
        } else {
            return new Shape(rawModel);
        }
    }

    protected static findShapeClass(shape: any, shapeClass: string): boolean {
        let types = shape["@type"];
        types = this.ensureArray(types);

        return types.find(t => t === shapeClass);
    }

    protected getValue<T>(p): T | undefined {
        let res = Shape.ensureArray(this.rawModel[p]);
        if (res.length > 0) {
            return res[0]["@value"] as T | undefined;
        } else {
            return undefined;
        }
    }

    protected getValues<T>(p): T[] | undefined {
        let res = Shape.ensureArray(this.rawModel[p]);
        if (res.length > 0) {
            return res.map(jsonld => jsonld["@value"]) as T[] | undefined;
        } else {
            return undefined;
        }
    }

    protected setValue(p: string, x: any) {
        if (x == null) {
            delete this.rawModel[p];
        } else {
            this.rawModel[p] = [{"@value": x}];
        }
    }

    protected setValues(p: string, x: any) {
        if (x == null) {
            delete this.rawModel[p];
        } else {
            this.rawModel[p] = Shape.ensureArray(x).map(x => [{"@value": x}]);
        }
    }

    protected setRdfObject(p: string, object: any | null) {
        if (object == null) {
            delete this.rawModel[p];
        } else {
            this.rawModel[p] = Shape.ensureArray(object);
        }
    }

    protected getRdfObject(p): any | undefined {
        let res = Shape.ensureArray(this.rawModel[p]);
        if (res.length > 0) {
            return res[0]
        }
    }

    protected getRdfObjects(p): any[] | undefined {
        return Shape.ensureArray(this.rawModel[p]);
    }

    protected getRdfId(p): string | undefined {
        let res = Shape.ensureArray(this.rawModel[p]);
        if (res.length > 0) {
            return res[0]["@id"]
        }
    }

    protected setRdfId(p: string, id: string | null) {
        if (id == null) {
            delete this.rawModel[p];
        } else {
            this.rawModel[p] = [{"@id": id}];
        }
    }

    protected static ensureArray(x: any): any[] {
        if (x == null) {
            return []
        } else if(x.constructor !== Array) {
            return [x];
        } else {
            return x;
        }
    }
}

import {NodeShape} from "./NodeShape";
import {ScalarShape} from "./ScalarShape";
import {PropertyShape} from "./PropertyShape";
import {ArrayShape} from "./ArrayShape";
import {FileUploadShape} from "./FileUploadShape";
import {XMLSchemaShape} from "./XMLSchemaShape";
import {JSONSchemaShape} from "./JSONSchemaShape";